metaflow/plugins/aws/utils.py

```diff
@@ -0,0 +1,192 @@
+            with tarfile.open(tar_file_path, 'r:gz') as tar:
+                tar.extractall(td)
+            copy_tree(
+                os.path.join(td, DATASTORE_LOCAL_DIR),
+                LocalDataStore.get_datastore_root_from_config(echo_none),
+                update=True)
+    except err as e:  # noqa F841
+        pass
+
+
+class CommonTaskAttrs:
+    def __init__(self,
+        flow_name,
+        step_name,
+        run_id,
+        task_id,
+        attempt,
+        user,
+        version,
+    ):
+        self.flow_name = flow_name
+        self.step_name = step_name
+        self.run_id = run_id
+        self.task_id = task_id
+        self.attempt = attempt
+        self.user = user
+        self.version = version
+
+    def to_dict(self, key_prefix):
+        attrs = {
+            key_prefix + 'flow_name': self.flow_name,
+            key_prefix + 'step_name': self.step_name,
+            key_prefix + 'run_id': self.run_id,
+            key_prefix + 'task_id': self.task_id,
+            key_prefix + 'retry_count': str(self.attempt),
+            key_prefix + 'version': self.version,
+        }
+        if self.user is not None:
+            attrs[key_prefix + 'user'] = self.user
+        return attrs
+
+def get_writeable_datastore(
+    common_task_attrs,
+    datastore_class,
+):
+    return datastore_class(
+            mode='w',
+            flow_name=common_task_attrs.flow_name,
+            step_name=common_task_attrs.step_name,
+            run_id=common_task_attrs.run_id,
+            task_id=common_task_attrs.task_id,
+            attempt=common_task_attrs.attempt
+    )
+
+def get_datastore_root(
+    common_task_attrs,
+    datastore_class,
+):
+    return datastore_class.make_path(common_task_attrs.flow_name, common_task_attrs.run_id, common_task_attrs.step_name, common_task_attrs.task_id)
+
+def build_task_preamble_commands(
+    environment,
+    code_package_url,
+    flow_name,
+    step_name,
+    run_id,
+    task_id,
+    attempt_str,
+    stdout_path,
+    stderr_path,
+):
+    mflog_expr = export_mflog_env_vars(datastore_type='s3',
+                                        stdout_path=stdout_path,
+                                        stderr_path=stderr_path,
+                                        flow_name=flow_name,
+                                        step_name=step_name,
+                                        run_id=run_id,
+                                        task_id=task_id,
+                                        retry_count=attempt_str,
+                                        datastore_root=None)
+    init_cmds = environment.get_package_commands(code_package_url)
+    init_cmds.extend(environment.bootstrap_commands(step_name))
+    init_expr = ' && '.join(init_cmds)
+
+    stdout_dir = os.path.dirname(stdout_path)
+    stderr_dir = os.path.dirname(stderr_path)
+
+    # Below we assume that stdout and stderr file are in the same (log) dir
+    assert stdout_dir == stderr_dir
```

Comment:
This is because we only do one mkdir call below. We could easily extend this code for the case when they are different directories, but in practice we always put these logs in one anyway. This assert is here just to make this assumption explicit.

Reply:
The trouble with assertion here is what do we expect the user to do next. Given that this assertion isn't supposed to fail any time in the future, we can probably do away with the assert statement.

Reply:
Right, but I wouldn't use the `assert` if there was any chance of this failing due to user input. _That_ should be a proper error, like throwing an exception or returning an error code. The idea here is to explicitly encode the assumptions made, so if some _future_ change violates that assumption, the code at least fails cleanly instead of causing some cryptic error far away from this place in code.

For example, in this case, if someone one day changes Metaflow so that stdout_dir is no longer the same as stderr_dir, tasks would fail (since the directory for stderr wouldn't exist). Without the assertion, it would still be a fatal error, but it would be harder to debug since all you'd get is some generic "can't open file" error from somewhere in cloudwatch logs.